fix handling of verify-attestation types for URIs

otms61 opened the pull request:
Signed-off-by: Akira Saso <sasoakira6114@gmail.com>

<!--
Thanks for opening a pull request! Please do not just delete this text.  The three fields below are mandatory.

Please remember to:
- This PR requires an issue. If it is a new feature, the issue should proceed the PR and will have allowed sufficent time for discussions to take place. Pleases use
issue tags such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ".
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!

Thank you :)
-->

#### Summary
<!--
 Explain the **motivation** for making this change. What existing problem does the pull request solve? How can reviewers test this PR?
-->

Close https://github.com/sigstore/cosign/issues/2158

Please feel free to close the PR if there is any requirement on verify-attestation.

I made an attestation with URI specified as the following command.
```bash
$ cat test-result.json
{"passed": true}

$ COSIGN_EXPERIMENTAL=1 cosign attest --type "https://example.com/TestResult/v1" --predicate test-result.json  otms61/test-custom-attest
```
Reviews can test this PR.

```bash
$ COSIGN_EXPERIMENTAL=1 cosign verify-attestation  --type "https://example.com/TestResult/v1" otms61/test-custom-attest
```

**before**
```bash
$ ./cosign-darwin-amd64 version
  ______   ______        _______. __    _______ .__   __.
 /      | /  __  \      /       ||  |  /  _____||  \ |  |
|  ,----'|  |  |  |    |   (----`|  | |  |  __  |   \|  |
|  |     |  |  |  |     \   \    |  | |  | |_ | |  . `  |
|  `----.|  `--'  | .----)   |   |  | |  |__| | |  |\   |
 \______| \______/  |_______/    |__|  \______| |__| \__|
cosign: A tool for Container Signing, Verification and Storage in an OCI registry.

GitVersion:    v1.10.1
GitCommit:     a39ce91fadc582e0efce3321744a79ccd3c8b39c
GitTreeState:  clean
BuildDate:     2022-08-04T16:59:14Z
GoVersion:     go1.18.5
Compiler:      gc
Platform:      darwin/amd64

$ COSIGN_EXPERIMENTAL=1 ./cosign-darwin-amd64 verify-attestation  --type "https://example.com/TestResult/v1" otms61/test-custom-attest
Error: converting to consumable policy validation: invalid predicate type: https://example.com/TestResult/v1
main.go:62: error during command execution: converting to consumable policy validation: invalid predicate type: https://example.com/TestResult/v1
```

**after**

```bash
$ COSIGN_EXPERIMENTAL=1 ./cosign verify-attestation  --type "https://example.com/TestResult/v1" otms61/test-custom-attest


Verification for otms61/test-custom-attest --
The following checks were performed on each of these signatures:
  - The cosign claims were validated
  - Existence of the claims in the transparency log was verified offline
  - Any certificates were verified against the Fulcio roots.
Certificate subject:  sasoakira6114@gmail.com
Certificate issuer URL:  https://accounts.google.com
{"payloadType":"application/vnd.in-toto+json","payload":"eyJfdHlwZSI6Imh0dHBzOi8vaW4tdG90by5pby9TdGF0ZW1lbnQvdjAuMSIsInByZWRpY2F0ZVR5cGUiOiJodHRwczovL2V4YW1wbGUuY29tL1Rlc3RSZXN1bHQvdjEiLCJzdWJqZWN0IjpbeyJuYW1lIjoiaW5kZXguZG9ja2VyLmlvL290bXM2MS90ZXN0LWN1c3RvbS1hdHRlc3QiLCJkaWdlc3QiOnsic2hhMjU2IjoiOTIyNTE0NTgwODhjNjM4MDYxY2RhOGZkOGI0MDNiNzZkNjYxYTRkYzZiN2VlNzFiNmFmZmNmMTg3MjU1N2IyYiJ9fV0sInByZWRpY2F0ZSI6eyJwYXNzZWQiOnRydWV9fQ==","signatures":[{"keyid":"","sig":"MEYCIQDq+COzci2fcmRHZk9P5Q9FzvfgNYJ7a/gvQbQo5tjNlQIhAMeflNARYm5HHc1QXUfayrlQVFOB7RvOjsGMl9vYMZQB"}]}
```


#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to an administrator running private sigstore instances (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->

#### Documentation
<!--

Does this change require an update to documentation? How will users implement your new feature?

Please reference a PR within https://docs.sigstore.dev

-->

otms61 committed:
fix handling of verify-attestation types for URIs

Signed-off-by: Akira Saso <sasoakira6114@gmail.com>

priyawadhwa reviewed (commented):
Looks good! Maybe we could add a quick test for this? https://github.com/sigstore/cosign/blob/main/test/e2e_test.go

otms61 commented:
Thanks for your review! I'll try.

otms61 committed:
test: add a test to verify URI-type attestation

Signed-off-by: Akira Saso <sasoakira6114@gmail.com>

otms61 committed:
test: fix cue used in verify-attstation

Signed-off-by: Akira Saso <sasoakira6114@gmail.com>

otms61 commented:
I have added a test to verify URI-type attestation.
And, I have made the following modifications.
- The verify-attestation command was not executed for badCue, so I made it execute.
- The value of goodCue was as (maybe) unintended and was passed, so I rewrote the value.
- I rewrote the value for badCue as well.

I am not familiar with CUE, but I noticed the following behavior.
As you can see in the example below, CUE was passing through the wrong path.

When the correct path and correct value, the command pass.
```
$ cat test-payload.json
{
  "predicate": {
    "Data": {
      "bomFormat": "CycloneDX",
      "specVersion": "1.4"
    }
  }
}

$ cat valid.cue
predicate: Data: specVersion: "1.4"

$ cue vet test-payload.json valid.cue

$
```

When the correct path and wrong value, the command don't pass.
```
$ cat invalid.cue
predicate: Data: specVersion: "7.4"

$ cue vet test-payload.json invalid.cue
predicate.Data.specVersion: conflicting values "7.4" and "1.4":
    ./invalid.cue:1:31
    ./test-payload.json:5:22

$
```

When the wrong path, the command pass.
```
$ cat current.cue
Data: specVersion: "7.4"

$ cue vet test-payload.json current.cue

$
```

The data path was wrong, so the test passed.
The following figure shows the json data, which CUE is applied when `TestAttestVerifyCycloneDXJSON`. You can see that the value is in predicate.Data.specVersion, not Data.specVersion.

![cue](https://user-images.githubusercontent.com/1850144/184798967-f2fda70e-c7d0-48b1-bc1d-0c9877c3f305.png)

<details> 

<summary> payload  </summary>

```json
{
  "_type": "https://in-toto.io/Statement/v0.1",
  "predicateType": "https://cyclonedx.org/schema",
  "subject": [
    {
      "name": "127.0.0.1:59647/cosign-attest-cyclonedx-e2e-image",
      "digest": {
        "sha256": "731a9ea8309ada1d7c57abb3cbe018c629e5202fe94c42d1a19bd70aaf6f4c8a"
      }
    }
  ],
  "predicate": {
    "Data": {
      "bomFormat": "CycloneDX",
      "components": [],
      "dependencies": [],
      "metadata": {},
      "serialNumber": "urn:uuid:9b0c2427-be94-439c-82e5-8928db124270",
      "specVersion": "1.4",
      "version": 1
    },
    "Timestamp": ""
  }
}
```

</details>
